File: 2021/java/src/Day1.java
import java.io.IOException;
import java.nio.file.Path;
import java.util.ArrayList;

public class Day1 implements Day {
    public static final String INPUT = "../inputs/input01";

    @Override
    public int part1() {
        var count = 0;
        try {
            var lines = Utilities.read_lines(Path.of(INPUT));
            var depths = lines.stream().map(Integer::parseInt).toList();
            for (var i = 1; i < depths.size(); ++i) {
                if (depths.get(i) > depths.get(i - 1)) count++;
            }
        } catch (IOException e) {
            System.err.println(e.getMessage());
        }
        return count;
    }

    @Override
    public int part2() {
        var count = 0;
        try {
            var lines = Utilities.read_lines(Path.of(INPUT));
            var depths = lines.stream().map(Integer::parseInt).toList();
            for (var i = 3; i < depths.size(); ++i) {
                if (depths.get(i) > depths.get(i - 3)) count++;
            }
        } catch (IOException e) {
            System.err.println(e.getMessage());
        }
        return count;
    }
}

File: 2021/cpp/day1.cpp
//
// Created by trox667 on 01.12.21.
//

#include <vector>
#include <string>

namespace day1 {
    int part1(const std::vector<std::string> &lines) {
        std::vector<int> depths(lines.size());
        std::transform(lines.cbegin(), lines.cend(), depths.begin(),
                       [](const std::string &line) {
                           return std::stoi(line);
                       });
        int count = 0;
        for(auto i = 1; i < depths.size(); ++i) {
            if (depths[i-1] < depths[i]) count++;
        }
        return count;
    }

    int part2(const std::vector<std::string> &lines) {
        std::vector<int> depths(lines.size());
        std::transform(lines.cbegin(), lines.cend(), depths.begin(),
                       [](const std::string &line) {
                           return std::stoi(line);
                       });
        int count = 0;
        for(auto i = 3; i < depths.size(); ++i) {
            if (depths[i-3] < depths[i]) count++;
        }
        return count;
    }
}